apis/s3/src/main/java/org/jclouds/s3/filters/RequestAuthorizeSignature.java

```diff
@@ -68,191 +68,192 @@
 /**
  * Signs the S3 request.
  */
-@Singleton
-public class RequestAuthorizeSignature implements HttpRequestFilter, RequestSigner {
 
-   private static final Collection<String> FIRST_HEADERS_TO_SIGN = ImmutableList.of(HttpHeaders.DATE);
+public interface RequestAuthorizeSignature extends HttpRequestFilter {
+    @Singleton
+    public static class RequestAuthorizeSignatureV2 implements RequestAuthorizeSignature, RequestSigner {
+        private static final Collection<String> FIRST_HEADERS_TO_SIGN = ImmutableList.of(HttpHeaders.DATE);
 
-   private static final Set<String> SIGNED_PARAMETERS = ImmutableSet.of("acl", "torrent", "logging", "location", "policy",
+        private static final Set<String> SIGNED_PARAMETERS = ImmutableSet.of("acl", "torrent", "logging", "location", "policy",
```

Comment:
Signature v4 need signed all of query string parameters.
_CanonicalQueryString specifies the URI-encoded query string parameters. You URI-encode name and values individually. You must also sort the parameters in the canonical query string alphabetically by key name. The sorting occurs after encoding._